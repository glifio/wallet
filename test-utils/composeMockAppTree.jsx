/* eslint-disable react/prop-types */
import React, { useReducer } from 'react'
import cloneDeep from 'lodash.clonedeep'

import WalletProvider from '@openworklabs/filecoin-wallet-provider'
import { FilecoinNumber } from '@openworklabs/filecoin-number'
import { Provider } from 'react-redux'

import { NetworkCheck } from '../lib/check-network'
import BalancePoller from '../lib/update-balance'
import { ConverterContext } from '../lib/Converter'
import { theme, ThemeProvider } from '../components/Shared'
import { initializeStore } from './index'
import { initialState } from '../store/states'
import walletProviderReducer, {
  initialState as walletProviderInitialState,
  setWalletType,
  setError,
  resetLedgerState,
  resetState
} from '../WalletProvider/state'
import { WalletProviderContext } from '../WalletProvider'
import createPath from '../utils/createPath'
import { IMPORT_MNEMONIC } from '../constants'

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

const mockRouter = jest.fn(() => {})
useRouter.mockImplementation(() => ({
  query: 'network=t',
  push: mockRouter,
  replace: mockRouter
}))

jest.mock('@openworklabs/filecoin-wallet-provider')
const mockGetAccounts = jest
  .fn()
  .mockImplementation(() =>
    Promise.resolve(['t1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq'])
  )

const mockGetBalance = jest
  .fn()
  .mockImplementation(() => Promise.resolve(new FilecoinNumber('1', 'fil')))

WalletProvider.mockImplementation(() => {
  return {
    wallet: {
      getAccounts: mockGetAccounts,
      sign: jest.fn().mockImplementation(() => 'xxxyyyyzzzz')
    },
    getBalance: mockGetBalance,
    getNonce: jest.fn().mockImplementation(() => 0),
    estimateGas: jest
      .fn()
      .mockImplementation(() => new FilecoinNumber('122', 'attofil')),
    sendMessage: jest.fn().mockImplementation(() => 'QmZCid!')
  }
})

const mockWalletProviderInstance = new WalletProvider()

const presets = {
  preOnboard: cloneDeep(initialState),
  postOnboard: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi',
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  })
}

const composeWalletProviderState = (initialWalletProviderState, preset) => {
  switch (preset) {
    case 'postOnboard': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance
      })
    }
    default:
      return initialWalletProviderState
  }
}

/**
 * VALID OPTIONS:
 * mockConverterInstance (an object that stubs the converter functionality)
 * walletProviderReducer (a reducer to stub the state handling of the wallet provider)
 * walletProviderInitialState(a reducer to stub the state of the walletProvider context)
 * pathname
 * query (usually { network: 'f' } or { network: 't' })
 */

export default (statePreset = 'preOnboard', options = {}) => {
  const state = options.state || presets[statePreset]
  const store = initializeStore(state)

  const mockConverterInstance = options.mockConverterInstance || {
    cacheConversionRate: jest.fn(),
    toFIL: jest.fn().mockImplementation(amount => {
      return new FilecoinNumber(amount, 'fil').dividedBy(5)
    }),
    fromFIL: jest.fn().mockImplementation(amount => {
      return new FilecoinNumber(amount, 'fil').multipliedBy(5)
    })
  }

  const pathname = options.pathname || '/wallet'
  const query = options.query || { network: 't' }

  const Tree = ({ children }) => {
    const [initialWalletProviderState, walletProviderDispatch] = useReducer(
      options.reducer || walletProviderReducer,
      options.walletProviderInitialState || walletProviderInitialState
    )

    const walletProviderState = composeWalletProviderState(
      initialWalletProviderState,
      statePreset
    )
    return (
      <Provider store={store}>
        <ConverterContext.Provider
          value={{
            converter: mockConverterInstance,
            converterError: options.converterError || null
          }}
        >
          <WalletProviderContext.Provider
            value={{
              state: walletProviderState,
              dispatch: walletProviderDispatch,
              fetchDefaultWallet: jest.fn().mockImplementation(),
              setWalletError: jest
                .fn()
                .mockImplementation(errorMessage =>
                  walletProviderDispatch(setError(errorMessage))
                ),
              setWalletType: jest
                .fn()
                .mockImplementation(walletType =>
                  walletProviderDispatch(setWalletType(walletType))
                ),
              setLedgerProvider: jest.fn(),
              connectLedger: jest.fn(),
              resetLedgerState: jest.fn().mockImplementation(() => {
                walletProviderDispatch(resetLedgerState())
              }),
              resetState: jest.fn().mockImplementation(() => {
                walletProviderDispatch(resetState())
              })
            }}
          >
            <NetworkCheck
              networkFromRdx={store.getState().network}
              pathname={pathname}
              query={query}
              switchNetwork={jest.fn()}
            />
            <BalancePoller />
            <ThemeProvider theme={theme}>{children}</ThemeProvider>
          </WalletProviderContext.Provider>
        </ConverterContext.Provider>
      </Provider>
    )
  }

  return { Tree, store, walletProvider: mockWalletProviderInstance }
}
