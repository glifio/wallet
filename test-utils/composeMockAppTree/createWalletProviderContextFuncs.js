import { FilecoinNumber } from '@glif/filecoin-number'

import {
  setWalletType,
  setError,
  resetLedgerState,
  resetState
} from '../../WalletProvider/state'

export default walletProviderDispatch => ({
  fetchDefaultWallet: jest.fn().mockImplementation(() => ({
    balance: new FilecoinNumber('1', 'fil'),
    address: 't1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq',
    path: "m/44'/1'/0/0/0"
  })),
  setWalletError: jest
    .fn()
    .mockImplementation(errorMessage =>
      walletProviderDispatch(setError(errorMessage))
    ),
  setWalletType: jest
    .fn()
    .mockImplementation(walletType =>
      walletProviderDispatch(setWalletType(walletType))
    ),
  setLedgerProvider: jest.fn(),
  connectLedger: jest.fn(),
  resetLedgerState: jest.fn().mockImplementation(() => {
    walletProviderDispatch(resetLedgerState())
  }),
  resetState: jest.fn().mockImplementation(() => {
    walletProviderDispatch(resetState())
  })
})
