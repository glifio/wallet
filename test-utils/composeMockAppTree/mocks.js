import WalletProvider from '@openworklabs/filecoin-wallet-provider'
import { FilecoinNumber } from '@openworklabs/filecoin-number'

import {
  setWalletType,
  setError,
  resetLedgerState,
  resetState
} from '../../WalletProvider/state'
import { SINGLE_KEY, HD_WALLET, LEDGER } from '../../constants'

jest.mock('@openworklabs/filecoin-wallet-provider')

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

export const mockRouterReplace = jest.fn(() => {})
export const mockRouterPush = jest.fn(() => {})
const unix = jest.spyOn(require('dayjs'), 'unix')

unix.mockImplementation(() => {
  return {
    format: fmt => {
      if (fmt === 'MMM DD') return 'Jan 12'
      if (fmt === 'hh:mmA') return '2:30PM'
    }
  }
})

useRouter.mockImplementation(() => ({
  query: 'network=t',
  push: mockRouterPush,
  replace: mockRouterReplace
}))

const mockGetAccounts = jest
  .fn()
  .mockImplementation(() =>
    Promise.resolve(['t1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq'])
  )

const mockGetBalance = jest
  .fn()
  .mockImplementation(() => Promise.resolve(new FilecoinNumber('1', 'fil')))

WalletProvider.mockImplementation(() => {
  return {
    wallet: {
      getAccounts: mockGetAccounts,
      sign: jest.fn().mockImplementation(() => 'xxxyyyyzzzz')
    },
    getBalance: mockGetBalance,
    getNonce: jest.fn().mockImplementation(() => 0),
    estimateGas: jest
      .fn()
      .mockImplementation(() => new FilecoinNumber('122', 'attofil')),
    sendMessage: jest.fn().mockImplementation(() => 'QmZCid!')
  }
})

export const mockWalletProviderInstance = new WalletProvider()

export const defaultMockConverterInstance = {
  cacheConversionRate: jest.fn(),
  toFIL: jest.fn().mockImplementation(amount => {
    return new FilecoinNumber(amount, 'fil').dividedBy(5)
  }),
  fromFIL: jest.fn().mockImplementation(amount => {
    return new FilecoinNumber(amount, 'fil').multipliedBy(5)
  })
}

export const createMockWalletProviderContextFuncs = walletProviderDispatch => ({
  fetchDefaultWallet: jest.fn().mockImplementation(() => ({
    balance: new FilecoinNumber('1', 'fil'),
    address: 't1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq',
    path: ''
  })),
  setWalletError: jest
    .fn()
    .mockImplementation(errorMessage =>
      walletProviderDispatch(setError(errorMessage))
    ),
  setWalletType: jest
    .fn()
    .mockImplementation(walletType =>
      walletProviderDispatch(setWalletType(walletType))
    ),
  setLedgerProvider: jest.fn(),
  connectLedger: jest.fn(),
  resetLedgerState: jest.fn().mockImplementation(() => {
    walletProviderDispatch(resetLedgerState())
  }),
  resetState: jest.fn().mockImplementation(() => {
    walletProviderDispatch(resetState())
  })
})

class MockSingleKeyProvider {
  type = SINGLE_KEY

  getAccounts = () => jest.fn()

  sign = () => jest.fn()
}

class MockHDWalletProvider {
  type = HD_WALLET

  getAccounts = () => jest.fn()

  sign = () => jest.fn()
}

export const mockWalletSubproviders = {
  SingleKeyProvider: MockSingleKeyProvider,
  HDWalletProvider: MockHDWalletProvider
}
