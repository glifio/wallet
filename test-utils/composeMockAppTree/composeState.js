import cloneDeep from 'lodash.clonedeep'
import { FilecoinNumber } from '@openworklabs/filecoin-number'

import createPath from '../../utils/createPath'
import { IMPORT_MNEMONIC } from '../../constants'
import { initialState } from '../../store/states'
import { mockWalletProviderInstance } from '../mocks/mock-wallet-provider'

export const presets = {
  preOnboard: cloneDeep(initialState),
  postOnboard: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi',
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  }),
  postOnboardWithError: cloneDeep({
    ...initialState,
    error: 'error for testing',
    wallets: [
      {
        address: 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi',
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  })
}

export const composeWalletProviderState = (
  initialWalletProviderState,
  preset
) => {
  switch (preset) {
    case 'postOnboard': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance
      })
    }
    case 'postOnboardWithError': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance
      })
    }
    default:
      return initialWalletProviderState
  }
}
