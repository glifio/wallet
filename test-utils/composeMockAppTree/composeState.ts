import cloneDeep from 'lodash.clonedeep'
import { FilecoinNumber } from '@glif/filecoin-number'

import createPath from '../../utils/createPath'
import { IMPORT_MNEMONIC } from '../../constants'
import { initialState } from '../../store/states'
import { mockWalletProviderInstance } from '../mocks/mock-wallet-provider'
import { emptyMsigState } from '../../MsigProvider/types'
import { WALLET_ADDRESS, MULTISIG_ACTOR_ADDRESS, signers } from '../constants'

export const presets = {
  preOnboard: cloneDeep(initialState),
  postOnboard: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: WALLET_ADDRESS,
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  }),
  postOnboardLowBal: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: WALLET_ADDRESS,
        balance: new FilecoinNumber('.000001', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  }),
  postOnboardWithError: cloneDeep({
    ...initialState,
    error: 'error for testing',
    wallets: [
      {
        address: WALLET_ADDRESS,
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  }),
  pendingMsigCreate: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: WALLET_ADDRESS,
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0,
    messages: {
      confirmed: [],
      pending: [
        {
          cid: 'bafy2bzaced2godar6dht6ag6omfa53jr6tykygtzskzodqo2k4ngvyou4yct2',
          from: 'f1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a',
          gasFeeCap: '100425',
          gasLimit: 11452096,
          gasPremium: '99371',
          maxFee: '1150076740800',
          method: 'EXEC',
          nonce: 56,
          paidFee: '0',
          params: {
            num_approvals_threshold: 1,
            signers: ['f1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a'],
            start_epoch: '0',
            unlock_duration: '0'
          },
          timestamp: 1632317830,
          to: 'f01'
        }
      ]
    }
  })
}

export const composeWalletProviderState = initialWalletProviderState => {
  return Object.freeze({
    ...initialWalletProviderState,
    walletType: IMPORT_MNEMONIC,
    walletProvider: mockWalletProviderInstance
  })
}

export const mockMsigProviderContext = {
  Address: MULTISIG_ACTOR_ADDRESS,
  ActorCode: 'fil/5/multisig',
  Balance: new FilecoinNumber('1', 'fil'),
  AvailableBalance: new FilecoinNumber('1', 'fil'),
  Signers: signers,
  InitialBalance: new FilecoinNumber('1', 'fil'),
  NextTxnID: 0,
  NumApprovalsThreshold: 1,
  StartEpoch: 0,
  UnlockDuration: 0,
  errors: {
    notMsigActor: false,
    connectedWalletNotMsigSigner: false,
    actorNotFound: false,
    unhandledError: ''
  },
  loading: false,
  setMsigActor: null
}

export const composeMsigProviderState = (preset: keyof typeof presets) => {
  switch (preset) {
    case 'preOnboard': {
      return Object.freeze(emptyMsigState)
    }
    case 'pendingMsigCreate': {
      return Object.freeze(emptyMsigState)
    }
    default: {
      return Object.freeze(mockMsigProviderContext)
    }
  }
}
