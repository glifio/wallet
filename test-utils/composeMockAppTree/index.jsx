/* eslint-disable react/prop-types */
import React, { useReducer } from 'react'
import { Provider } from 'react-redux'
import { Converter } from '@openworklabs/filecoin-number'

import { NetworkCheck } from '../../lib/check-network'
import BalancePoller from '../../lib/update-balance'
import { ConverterContext } from '../../lib/Converter'
import { theme, ThemeProvider } from '../../components/Shared'
import { initializeStore } from '..'
import walletProviderReducer, {
  initialState as walletProviderInitialState
} from '../../WalletProvider/state'
import { WasmContext } from '../../lib/WasmLoader'
import { WalletProviderContext } from '../../WalletProvider'
import { mockWalletProviderInstance } from '../mocks/mock-wallet-provider'
import createMockWalletProviderContextFuncs from './createWalletProviderContextFuncs.js'
import mockWalletSubproviders from '../mocks/mock-wallet-subproviders'
import { presets, composeWalletProviderState } from './composeState'

/**
 * VALID OPTIONS:
 * mockConverterInstance (an object that stubs the converter functionality)
 * walletProviderReducer (a reducer to stub the state handling of the wallet provider)
 * walletProviderInitialState(a reducer to stub the state of the walletProvider context)
 * walletProviderDispatch(a dispatcher to mock/read calls to the walletProvider's state)
 * pathname
 * query (usually { network: 'f' } or { network: 't' })
 */

export default (statePreset = 'preOnboard', options = {}) => {
  const state = options.state || presets[statePreset]
  const store = initializeStore(state)

  const pathname = options.pathname || '/wallet'
  const query = options.query || { network: 't' }

  const Tree = ({ children }) => {
    const [initialWalletProviderState, walletProviderDispatch] = useReducer(
      options.reducer || walletProviderReducer,
      options.walletProviderInitialState || walletProviderInitialState
    )

    const walletProviderState = composeWalletProviderState(
      initialWalletProviderState,
      statePreset
    )

    const mockWalletProviderContextFuncs = createMockWalletProviderContextFuncs(
      options.walletProviderDispatch || walletProviderDispatch
    )

    return (
      <Provider store={store}>
        <WasmContext.Provider
          value={{
            generateMnemonic: jest
              .fn()
              .mockImplementation(
                () =>
                  'slender spread awkward chicken noise useful thank dentist tip bronze ritual explain version spot collect whisper glow peanut bus local country album punch frown'
              )
          }}
        >
          <ConverterContext.Provider
            value={{
              converter: new Converter(),
              converterError: options.converterError || null
            }}
          >
            <WalletProviderContext.Provider
              value={{
                state: walletProviderState,
                dispatch:
                  options.walletProviderDispatch || walletProviderDispatch,
                walletSubproviders: {
                  ...mockWalletSubproviders
                },
                ...mockWalletProviderContextFuncs
              }}
            >
              <NetworkCheck
                networkFromRdx={store.getState().network}
                pathname={pathname}
                query={query}
                switchNetwork={jest.fn()}
              />
              <BalancePoller />
              <ThemeProvider theme={theme}>{children}</ThemeProvider>
            </WalletProviderContext.Provider>
          </ConverterContext.Provider>
        </WasmContext.Provider>
      </Provider>
    )
  }

  return {
    Tree,
    store,
    walletProvider: mockWalletProviderInstance
  }
}
