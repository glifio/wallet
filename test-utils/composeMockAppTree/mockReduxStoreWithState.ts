import { presets } from './composeState'
import { initializeStore } from '..'

export default function mockReduxStoreWithState(options: {
  state?: object
  statePreset?: string
}) {
  const state = options.state || presets[options.statePreset]
  return initializeStore(state)
}
