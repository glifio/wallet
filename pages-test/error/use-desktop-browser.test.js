import { cleanup, render, screen, act, fireEvent } from '@testing-library/react'

import UseDesktopBrowser from '../../pages/error/use-desktop-browser.jsx'
import composeMockAppTree from '../../test-utils/composeMockAppTree'

jest.mock('@openworklabs/filecoin-wallet-provider')

describe('UseDesktopBrowser', () => {
  afterEach(cleanup)
  test('it renders the error page', () => {
    const { Tree } = composeMockAppTree('postOnboard')
    const { container } = render(
      <Tree>
        <UseDesktopBrowser />
      </Tree>
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  test('it renders the message about getting Chrome', () => {
    const { Tree } = composeMockAppTree('postOnboard')

    render(
      <Tree>
        <UseDesktopBrowser />
      </Tree>
    )

    expect(
      screen.getByText('Glif Wallet isnâ€™t ready for your phone or tablet.')
    ).toBeInTheDocument()
  })

  test('it renders the home page after clicking back', async () => {
    const { Tree } = composeMockAppTree('preOnboard')
    const useRouter = jest.spyOn(require('next/router'), 'useRouter')
    const mockRouterReplace = jest.fn(() => {})
    useRouter.mockImplementationOnce(() => ({
      query: 'network=t',
      push: mockRouterReplace
    }))

    let res
    await act(async () => {
      res = render(
        <Tree>
          <UseDesktopBrowser />
        </Tree>
      )
      fireEvent.click(screen.getByText('Home'))
    })

    expect(mockRouterReplace).toHaveBeenCalled()
    expect(mockRouterReplace).toHaveBeenCalledWith('/')
  })
})
