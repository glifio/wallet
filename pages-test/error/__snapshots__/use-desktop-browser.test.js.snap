// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UseDesktopBrowser it renders the error page 1`] = `
.c0 {
  min-width: 0;
  box-sizing: border-box;
  padding: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 100vh;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  min-width: 0;
  box-sizing: border-box;
  padding-left: 32px;
  padding-right: 32px;
  text-align: center;
}

.c2 {
  min-width: 0;
  box-sizing: border-box;
  border-radius: 8px;
  border: 1px solid;
  border-color: #FC6D6F;
  margin: 0 auto;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 8px;
  padding-right: 8px;
  background-color: #FC6D6F;
  display: inline-block;
}

.c6 {
  cursor: pointer;
  border: 1px solid #262626;
  background-color: transparent;
  border-color: #262626;
  color: #262626;
  font-size: 1.125rem;
  -webkit-transition: 0.18s ease-in-out;
  transition: 0.18s ease-in-out;
  border-radius: 4px;
  margin-top: 32px;
  padding: 16px;
  border-width: 1px;
  border-radius: 4px;
}

.c6:hover {
  opacity: 0.8;
}

.c3 {
  width: 64px;
  height: 64px;
}

.c4 {
  font-size: 3rem;
  font-weight: 700;
  line-height: 1;
  font-family: RT-Alias-Medium;
  margin: 0;
  margin-top: 32px;
  margin-bottom: 32px;
}

.c5 {
  font-size: 1.5rem;
  font-weight: 400;
  line-height: 1.2;
  font-family: RT-Alias-Grotesk;
  text-align: left;
  margin: 0;
  margin-top: 32px;
  margin-bottom: 32px;
}

<div
  class="c0"
  display="flex"
  width="100%"
>
  <div
    class="c1"
  >
    <div
      class="c2"
      display="inline-block"
    >
      <svg
        class="c3"
        fill="status.fail.foreground"
        height="24"
        size="7"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M7.28131 9.18344C7.2736 9.16518 7.2696 9.14561 7.26953 9.12582V1.15017C7.26972 1.11023 7.28591 1.072 7.31454 1.04385C7.34316 1.01571 7.3819 0.999936 7.42226 1H19.8441C19.8844 0.999936 19.9232 1.01571 19.9518 1.04385C19.9804 1.072 19.9966 1.11023 19.9968 1.15017V2.64896C19.9967 2.66874 19.9927 2.68832 19.985 2.70658C19.9773 2.72483 19.966 2.7414 19.9519 2.75534C19.9377 2.76929 19.9209 2.78033 19.9024 2.78785C19.8839 2.79536 19.8641 2.79919 19.8441 2.79913H9.81499C9.6221 2.79913 9.43712 2.87495 9.30073 3.00991C9.16434 3.14487 9.08771 3.32792 9.08771 3.51878V6.75721C9.08771 6.94807 9.16434 7.13112 9.30073 7.26608C9.43712 7.40104 9.6221 7.47686 9.81499 7.47686H17.4513C17.6442 7.47686 17.8292 7.40104 17.9656 7.26608C18.102 7.13112 18.1786 6.94807 18.1786 6.75721C18.1786 6.56635 18.102 6.3833 17.9656 6.24834C17.8292 6.11338 17.6442 6.03756 17.4513 6.03756H10.695C10.675 6.03762 10.6552 6.03379 10.6367 6.02628C10.6182 6.01876 10.6014 6.00772 10.5872 5.99378C10.573 5.97983 10.5617 5.96326 10.554 5.94501C10.5463 5.92675 10.5423 5.90718 10.5423 5.88739V4.3886C10.5423 4.36882 10.5463 4.34924 10.554 4.33098C10.5617 4.31273 10.573 4.29616 10.5872 4.28222C10.6014 4.26827 10.6182 4.25723 10.6367 4.24971C10.6552 4.2422 10.675 4.23837 10.695 4.23843H19.8441C19.8641 4.23837 19.8839 4.2422 19.9024 4.24971C19.9209 4.25723 19.9377 4.26827 19.9519 4.28222C19.966 4.29616 19.9773 4.31273 19.985 4.33098C19.9927 4.34924 19.9967 4.36882 19.9968 4.3886V9.12582C19.9967 9.14561 19.9927 9.16518 19.985 9.18344C19.9773 9.20169 19.966 9.21826 19.9519 9.23221C19.9377 9.24615 19.9209 9.25719 19.9024 9.26471C19.8839 9.27222 19.8641 9.27605 19.8441 9.27599H7.42226C7.40227 9.27605 7.38246 9.27222 7.36396 9.26471C7.34547 9.25719 7.32865 9.24615 7.31447 9.23221C7.30028 9.21826 7.28902 9.20169 7.28131 9.18344ZM19.9549 15.7578C19.9262 15.7859 19.8874 15.8016 19.847 15.8015H4.15297C4.13286 15.8016 4.11294 15.7977 4.09436 15.7901C4.07577 15.7825 4.05888 15.7714 4.04466 15.7573C4.03045 15.7433 4.01918 15.7266 4.01152 15.7082C4.00385 15.6898 3.99994 15.6701 4 15.6502V14.1524C4.00019 14.1125 4.01641 14.0742 4.04509 14.0461C4.07377 14.018 4.11257 14.0022 4.15297 14.0024H19.847C19.8874 14.0022 19.9262 14.018 19.9549 14.0461C19.9836 14.0742 19.9998 14.1125 20 14.1524V15.6514C19.9998 15.6914 19.9836 15.7296 19.9549 15.7578ZM19.847 12.5603C19.867 12.5604 19.8868 12.5565 19.9053 12.5491C19.9238 12.5416 19.9407 12.5306 19.9549 12.5166C19.9691 12.5027 19.9804 12.4862 19.9881 12.4679C19.9959 12.4497 19.9999 12.4301 20 12.4103V10.9113C19.9998 10.8713 19.9836 10.8331 19.9549 10.8049C19.9262 10.7767 19.8874 10.761 19.847 10.7611H6.54546C6.35257 10.7611 6.16759 10.6853 6.0312 10.5503C5.89481 10.4154 5.81818 10.2323 5.81818 10.0415V1.1972C5.81831 1.15722 5.8024 1.11883 5.77395 1.09045C5.7455 1.06207 5.70682 1.04602 5.66643 1.04583H4.15176C4.11136 1.04602 4.07269 1.06207 4.04423 1.09045C4.01578 1.11883 3.99987 1.15722 4 1.1972V12.4103C4.00001 12.4485 4.01485 12.4852 4.04146 12.5128L4.04485 12.5164C4.0588 12.5301 4.07536 12.5411 4.09358 12.5485C4.11179 12.556 4.13132 12.5598 4.15103 12.5598H19.847V12.5603ZM4.01106 22.9071C4.00354 22.8886 3.99979 22.8689 4.00001 22.849V17.3945C4.0002 17.3545 4.01642 17.3162 4.0451 17.2881C4.07378 17.2599 4.11258 17.2442 4.15298 17.2443H19.847C19.8874 17.2442 19.9262 17.2599 19.9549 17.2881C19.9836 17.3162 19.9998 17.3545 20 17.3945V18.8929C19.9998 18.9329 19.9836 18.9712 19.9549 18.9993C19.9262 19.0275 19.8874 19.0432 19.847 19.0431H6.54546C6.35258 19.0431 6.16759 19.1189 6.0312 19.2538C5.89481 19.3887 5.81819 19.5718 5.81819 19.7626C5.81819 19.9534 5.89481 20.1364 6.0312 20.2713C6.16759 20.4063 6.35258 20.4821 6.54546 20.4821H19.847C19.8874 20.482 19.9262 20.4977 19.9549 20.5258C19.9836 20.554 19.9998 20.5922 20 20.6322V22.1319C19.9998 22.1719 19.9836 22.2101 19.9549 22.2383C19.9262 22.2664 19.8874 22.2822 19.847 22.282H6.39249C6.24039 22.282 6.0945 22.3417 5.98683 22.448C5.87916 22.5543 5.81851 22.6985 5.81819 22.849C5.81842 22.8689 5.81466 22.8886 5.80714 22.9071C5.79963 22.9255 5.78849 22.9423 5.77439 22.9565C5.76029 22.9707 5.74349 22.982 5.72496 22.9897C5.70643 22.9974 5.68654 23.0014 5.66643 23.0015H4.15177C4.13166 23.0014 4.11177 22.9974 4.09324 22.9897C4.07472 22.982 4.05792 22.9707 4.04381 22.9565C4.02971 22.9423 4.01858 22.9255 4.01106 22.9071Z"
          fill="status.fail.foreground"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <h1
      class="c4"
      font-family="RT-Alias-Medium"
      font-size="6"
      font-weight="700"
    >
      Not yet
    </h1>
    <h2
      class="c5"
      font-family="RT-Alias-Grotesk"
      font-size="4"
      font-weight="400"
    >
      Glif Wallet isnâ€™t ready for your phone or tablet.
    </h2>
    <h2
      class="c5"
      font-family="RT-Alias-Grotesk"
      font-size="4"
      font-weight="400"
    >
      Please access it from your computer instead.
    </h2>
    <button
      class="c6"
      font-size="3"
      type="button"
    >
      Home
    </button>
  </div>
</div>
`;
